sw/device/silicon_creator/rom/doc/bootstrap.md

```diff
@@ -0,0 +1,189 @@
+```
+
+### Difficulties Anticipated
+
+#### Violation of SPI Expectations
+
+We may not be able to implement the READ opcode due to design limitations in the SPI device or due to code confidentiality requirements for certification.
+In such a scenario, the SPI programmer must be configured to issue the programming sequence without performing a READ/VERIFY sequence.
+This is usually a normal programmer configuration option.
+
+The SPI RESET opcode normally requires a sequence of RSTEN followed by RESET.
+Normally RSTEN followed by any other opcode cancels the reset enable.
+I’m not expecting to be able to implement enforcement of this sequencing because intercepting RSTEN in software will create latency for handling the next command.
+
+#### Debug and Upstreaming patches for Dediprog Software
+
+The open source Dediprog SF-100 software is very low-quality code.
+Although it is a goal to work with the Dediprog (and other commodity SPI programmer) devices without modifying the software, we must anticipate that the software will require modifications.
+
+At a minimum, I expect that we’ll need to upstream a fragment of XML configuration for the Dediprog “chipdb” which is used to inform the software about the properties of the target device.
+
+## Test Plan
+
+The bootstrap module will use unit tests to validate the proper operation at the module boundaries.
+
+The bootstrap module will also have functional tests which use verilator to simulate the bootstrap flows via the SPI device interface.
+These tests will check for valid operation of the bootstrap flows and will check that the properties of each bootstrap phase perform their intended functions and correctly abort when the protocol sequence is violated.
+
+Because the bootstrap module implements the real-world programming interface, testing flows must be developed which verify bootstrap functionality on FPGA devices paired with real programmer devices such as the Dediprog SF100.
+Any difficulties encountered with the physical devices should result in enhanced tests performed in the unit tests or function tests.
+
+## Alternatives Considered
+
+The pre-existing bootstrap protocol used on current Google Titan-Class chips is well known (at least within Google) and has existing library support both in Google's codebase and in Open Source repositories (i.e.: the `spiflash` utility).
+
+Although this protocol is well known and has existing support, this protocol requires the device driving the bootstrap sequence support SPI in generic mode and be a full-duplex SPI transceiver.
+
+The author has on very rare occasions observed peculiar problems with the exiting bootstrap protocol which usually present as corrupted SHA checksums (usually in the form of a dropped bit).
+These can normally be worked around by ignoring the checksums.
+
+There are also reports of occasional synchronization issues with the full-duplex protocol.
+
+[^1]: E.g. XMODEM or parsing of S-records.
+[^2]: E.g. The SPI EEPROM protocol.
```

Comment:
These footnotes are broken in the site render and don't necessarily need to be footnotes.

I am going to propose some updates to the markdown style guide soon. One of these is to suggest we stick to [CommonMark](https://spec.commonmark.org/) with the [GFM Tables](https://github.github.com/gfm/#tables-extension-) and the [GFM List Items](https://github.github.com/gfm/#task-list-items-extension-) extensions only. I'm planning to leave out footnotes because there exists no standard and agreed upon way to handle footnotes in markdown.